openmls/src/key_packages/key_package_in.rs

```diff
@@ -0,0 +1,196 @@
+#[derive(
+    Debug, Clone, PartialEq, TlsSize, TlsSerialize, TlsDeserialize, Serialize, Deserialize,
+)]
+struct KeyPackageTbsIn {
+    protocol_version: ProtocolVersion,
+    ciphersuite: Ciphersuite,
+    init_key: HpkePublicKey,
+    leaf_node: LeafNodeIn,
+    extensions: Extensions,
+}
+
+/// The key package struct.
+#[derive(
+    Debug, PartialEq, Clone, Serialize, Deserialize, TlsSerialize, TlsDeserialize, TlsSize,
+)]
+pub struct KeyPackageIn {
+    payload: KeyPackageTbsIn,
+    signature: Signature,
+}
+
+impl KeyPackageIn {
+    /// Returns a [`CredentialWithKey`] from the unverified payload
+    pub fn unverified_credential(&self) -> CredentialWithKey {
+        let credential = self.payload.leaf_node.credential().clone();
+        let signature_key = self.payload.leaf_node.signature_key().clone();
+        CredentialWithKey {
+            credential,
+            signature_key,
+        }
+    }
+
+    /// Verify that this key package is valid:
+    /// * verify that the signature on this key package is valid
+    /// * verify that the signature on the leaf node is valid
+    /// * verify that all extensions are supported by the leaf node
+    /// * make sure that the lifetime is valid Returns a [`KeyPackage`] after
+    /// having verified the signature or a [`KeyPackageVerifyError`] otherwise.
+    pub fn into_validated(
+        self,
+        crypto: &impl OpenMlsCrypto,
+        ciphersuite: Ciphersuite,
+    ) -> Result<KeyPackage, KeyPackageVerifyError> {
+        // Verify the signature of the KeyPackage
+        let signature_key = OpenMlsSignaturePublicKey::from_signature_key(
+            self.payload.leaf_node.signature_key().clone(),
+            self.payload.ciphersuite.signature_algorithm(),
+        );
+        self.verify_no_out(crypto, &signature_key)
+            .map_err(|_| KeyPackageVerifyError::InvalidSignature)?;
+
+        // We need to verify the LeafNode inside the KeyPackage
+        let leaf_node = self.payload.leaf_node.into_verifiable_leaf_node();
+
+        let leaf_node = match leaf_node {
+            VerifiableLeafNode::KeyPackage(leaf_node) => {
+                let pk = &leaf_node
+                    .signature_key()
+                    .clone()
+                    .into_signature_public_key_enriched(ciphersuite.signature_algorithm());
+
+                leaf_node
+                    .verify(crypto, pk)
+                    .map_err(|_| KeyPackageVerifyError::InvalidLeafNodeSignature)?
+            }
+            _ => return Err(KeyPackageVerifyError::InvalidLeafNodeSourceType),
+        };
+
+        let key_package = KeyPackage {
+            payload: KeyPackageTBS {
+                protocol_version: self.payload.protocol_version,
+                ciphersuite: self.payload.ciphersuite,
+                init_key: self.payload.init_key,
+                leaf_node,
+                extensions: self.payload.extensions,
+            },
+            signature: self.signature,
+        };
```

Comment:
I'm trying to figure out what a good pattern would be here. I would expect `KeyPackageTbsIn` to implement `into_validated` as well, which then validates its own fields (extensions, leaf node). I'm not sure what should happen next, but it would be great if we could call `verify` instead of `verify_no_out` on the KeyPackage in the end and just return the result.

One way to do this would be to build a `VerifiableKeyPackage` from the (validated) KeyPackageTbs and then validate that. That would be a little cleaner, but would also require another struct.

Reply:
The current pattern just doesn't work very well for this case. I figured it was ok to do it like that since all of it happens within the `into_validated()` function.